protocols/ML01.md

```diff
@@ -0,0 +1,315 @@
+
+```javascript
+{
+    "status": alternative([
+        "running", "complete", "failed", "cancelled"
+    ]),
+
+    "job_type_id": string,
+    "started_by": string,
+    
+    "messages": optional(array({
+        "type": alternative([
+            "info", "warning", "error",
+        ]),
+        "time": string,
+        "message": string
+    }))
+
+    "started": string,
+    "completed": optional(string),
+    
+    "progress_estimate": optional(number)
+}
+```
+
+"started_by" give the username of the user who started the job.
+
+"messages" gives log messages output by the job as it ran, time is an ISO 8601 datetime given in UTC.
+
+started and completed are datetimes formatted according to ISO 8601 in UTC. If the job is still in progress it **MUST NOT** have a "completed" attribute. Progress estimate gives a normalised value from 0 to 1 estimating the percentage complete of the job.
+
+# Listing All Jobs
+
+To get the status of all running jobs on a given project the client should
+request
+
+```
+http://ml.endpoint/projects/<project_name>/jobs
+```
+
+If the given project does not exist the server **SHOULD** return error code 404 and error "project_not_found".
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned.
+
+Otherwise the server's unwrapped response **MUST** match
+```javascript
+array({
+    "job_id": string,
+    "status": {}
+})
+```
+
+where "status" gives the object defined above describing the jobs current status.
+
+# Stopping a Job
+
+To stop a job the client should POST to
+
+```
+http://ml.endpoint/projects/<project_name>/jobs/<job_id>?action=cancel
+```
+
+If the given project does not exist the server **SHOULD** return error code 404 and error "project_not_found".
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned.
+
+If the given job does not exist then the server **SHOULD** return error code 404 and error "job_not_found".
+
+If the job cannot be cancelled because it is not running the server **SHOULD** return error code 400 and error "invalid_job_state".
+
+otherwise the server **MUST** return a successful empty response.
+
+# Deleting a Job.
+
+To delete a job the client should POST to
+
+```
+http://ml.endpoint/projects/<project_name>/jobs/<job_id>?action=delete
+```
+
+If the given project does not exist the server **SHOULD** return error code 404 and error "project_not_found".
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned.
+
+If the given job does not exist then the server **SHOULD** return error code 404 and error "job_not_found".
+
+otherwise the server **MUST** return a successful empty response.
+
+# Notes for implementors
+## State
+This specification and BE01 were together written to allow implementations that did not depend on the ML endpoints requiring any storage. As an example of how this may work:
```

Comment:
It's probably just me being an idiot, but I didn't understand this bit. Could you ELI5?

Reply:
Idea is that ML nodes **could** store all working data in backend without need for any local storage / databases.  This is nothing new - I believe it was you who originally proposed the idea of backend storing MLs models, this is just a restatement of that.

Reply:
Ah okay, I was confused because earlier you'd said there is no need BE <-> ML communication, when this would require ML making calls to BE, no?